### PR TITLE
Completion list show correct entry for function expression and class expression

### LINKED ISSUE #3231: Named function expression bodies don't have completion info for their names
image

### PR DESCRIPTION
**Pull request by yuit**
Fix #3231 and for class expression case

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -2807,6 +2810,22 @@ namespace ts {
...
                 }
 
+                // Special case for function expression and class expression because despite sometimes having a name, the binder
+                // binds them to a symbol with the name "__function" and "__class" respectively. However, for completion entry, we want
+                // to display its declared name rather than "__function" and "__class".
+                //      var x = function foo () {
+                //          fo$  <- completion list should contain local name "foo"
+                //      }
+                //      foo$ <- completion list should not contain "foo"
+                if (displayName === "__function" || displayName === "__class") {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Look into `getDeclaredName` for logic, including the above for the `default` export

Comment (yuit) [inline src/services/services.ts]: ```getDeclaredName``` doesn't handle function expression correctly. Talk with @mhegazy, the function needed some rewrite and I don't want to have it done in this PR. I will put a TODO here to have it change once ```getDeclaredName``` is done

File: src/services/services.ts [hunk 2]
```diff
@@ -1423,6 +1423,9 @@ namespace ts {
         // class X {}
         export const classElement = "class";
 
+        // var x = class X {}
+        export const localClassElement = "local class";
+
```
Comment (mhegazy) [inline src/services/services.ts]: note. you need to handle this on the managed side as well.

File: src/services/services.ts (hunk 1 above)
Comment (mhegazy) [inline src/services/services.ts]: @DanielRosenwasser should have a fix for this in a pending review. you should synchronize.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -12170,11 +12170,6 @@ namespace ts {
...
                             copySymbols(getSymbolOfNode(location).exports, meaning & SymbolFlags.EnumMember);
                             break;
-                        case SyntaxKind.ClassExpression:
-                            if ((<ClassExpression>location).name) {
-                                copySymbol(location.symbol, meaning);
-                            }
-                            // Fall through
                         case SyntaxKind.ClassDeclaration:
                         case SyntaxKind.InterfaceDeclaration:
                             if (!(memberFlags & NodeFlags.Static)) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: what does line do?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -12182,8 +12177,13 @@ namespace ts {
                             }
                             break;
                         case SyntaxKind.FunctionExpression:
-                            if ((<FunctionExpression>location).name) {
-                                copySymbol(location.symbol, meaning);
+                        case SyntaxKind.ClassExpression:
+                            let name = (<FunctionExpression|ClassExpression>location).name;
+                            if (name) {
+                                let symbol = location.symbol;
+                                if (symbol.flags & meaning && !hasProperty(symbols, name.text)) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: i would add a comment why we are not using copySymbol.. cause i will forget :)

File: src/services/services.ts [hunk 5]
```diff
@@ -3761,7 +3781,16 @@ namespace ts {
                 }
             }
             if (symbolFlags & SymbolFlags.Class && !hasAddedSymbolInfo) {
-                displayParts.push(keywordPart(SyntaxKind.ClassKeyword));
+                // Special case for class expressions because we would like to indicate that
+                // the class name is local to the class body (similar to function expression)
+                //      (local class) class <className>
+                if (symbol.getName() === "__class") {
```
Comment (mhegazy) [inline src/services/services.ts]: why not use `getDeclarationOfKind(symbol, SyntaxKind.ClassExpression)` instead

File: src/compiler/checker.ts (hunk 3 above)
Comment (yuit) [inline src/compiler/checker.ts]: which line ?

File: src/compiler/checker.ts (hunk 4 above)
Comment (yuit) [inline src/compiler/checker.ts]: That is a good point :)

File: src/services/services.ts (hunk 2 above)
Comment (yuit) [inline src/services/services.ts]: What you mean be handling in the managed side?

File: src/services/services.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: Ah right, it's in my PR. Speaking of which, can you guys review it? :smiley: 

### #3367

File: src/compiler/checker.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: the line below, GitHub doesn't show it

File: src/services/services.ts [hunk 6]
```diff
@@ -3552,7 +3571,8 @@ namespace ts {
         function getSymbolKind(symbol: Symbol, location: Node): string {
             let flags = symbol.getFlags();
 
-            if (flags & SymbolFlags.Class) return ScriptElementKind.classElement;
+            if (flags & SymbolFlags.Class) return symbol.declarations[0].kind === SyntaxKind.ClassExpression ?
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Use getDeclarationOfKind

**Comment by DanielRosenwasser** [comment]
@yuit if you pull in from `master`, there's a fourslash test called `renameLocationsForClassExpression01.ts`. You'll need to amend it to test correctly in light of the fact that you fixed class expression name resolution.

File: src/services/services.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: Probably best to discuss this in person with someone but I believe the idea is that VS needs to handle styles differently depending on what comes in to the managed side.

File: src/services/services.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: You should be clear for `getDeclaredName`

Comment (yuit) [inline src/services/services.ts]: :rose: yeah

File: src/compiler/checker.ts (hunk 4 above)
Comment (yuit) [inline src/compiler/checker.ts]: :cactus:

File: src/services/services.ts (hunk 5 above)
Comment (yuit) [inline src/services/services.ts]: :cactus:

File: src/services/services.ts (hunk 6 above)
Comment (yuit) [inline src/services/services.ts]: :cactus:

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -12684,8 +12679,17 @@ namespace ts {
                             }
                             break;
                         case SyntaxKind.FunctionExpression:
-                            if ((<FunctionExpression>location).name) {
-                                copySymbol(location.symbol, meaning);
+                        case SyntaxKind.ClassExpression:
+                            // The reason we are not using copySymbol for function expression and class expression
+                            // is that copySymbol will not copy a symbol into SymbolTable if the symbol has name prefix
+                            // with "__". Therefore, if class expression or function expression have declared name,
+                            // we will add the symbol into the table using its declared name
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Why not change the logic of `copySymbol`?

Also, the comment for `copySymbol` says `// Returns 'true' if we should stop processing symbols.`, but it returns `void`.

File: src/harness/fourslash.ts [hunk 8]
```diff
@@ -704,13 +704,55 @@ module FourSlash {
             }
         }
 
-        public verifyCompletionListDoesNotContain(symbol: string) {
+        /**
+         * Verfiy that the completion list does NOT contain the given symbol. If text, or documentation, or kind are provided,
+         * the list contains the symbol all given parameters must matched. When any parameter is omitted, the parameters is ignored during comparison.
+         */
+        public verifyCompletionListDoesNotContain(symbol: string, expectedText?: string, expectedDocumentation?: string, expectedKind?: string) {
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Why do any of these parameters help? You're interested n the lack of `symbol` in the list, so if `symbol` isn't in the list, then why do you care about any of the other parameters.

File: src/services/services.ts [hunk 9]
```diff
@@ -2797,18 +2800,30 @@ namespace ts {
...
-                        displayName = symbol.valueDeclaration.localSymbol.name;
-                    }
-                }
-
+        /**
+         * Get the name to be display in completion from a given symbol.
+         *
+         * @return undefined if the name is of external module otherwise a name with striped of any quote
+         */
+        function getCompletionEntryDisplayNameForSymbol(symbol: Symbol, target: ScriptTarget, performCharacterChecks: boolean, location: Node): string {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Can you not just use `getDeclaredName`?

File: src/services/services.ts (hunk 2 above)
Comment (mhegazy) [inline src/services/services.ts]: what do you see in navbar when you do this?

File: src/services/services.ts [hunk 10]
```diff
@@ -3813,7 +3829,16 @@ namespace ts {
                 }
             }
             if (symbolFlags & SymbolFlags.Class && !hasAddedSymbolInfo) {
-                displayParts.push(keywordPart(SyntaxKind.ClassKeyword));
+                // Special case for class expressions because we would like to indicate that
+                // the class name is local to the class body (similar to function expression)
+                //      (local class) class <className>
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Move into the `if` body

File: src/services/services.ts [hunk 11]
```diff
@@ -2821,37 +2836,37 @@ namespace ts {
...
-                // If the user entered name for the symbol was quoted, removing the quotes is not enough, as the name could be an
-                // invalid identifier name. We need to check if whatever was inside the quotes is actually a valid identifier name.
-                displayName = displayName.substring(1, displayName.length - 1);
-            }
+            name = stripQuotes(name);
 
-            if (!displayName) {
+            // We can simply return name with strip quotes because the name could be an invalid identifier name
+            // e.g "b a" is valid quoted name but when we strip off the quotes, it is invalid.
+            // We, thus, need to check if whatever was inside the quotes is actually a valid identifier name.
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: It doesn't look like this is in the right place. I'd put it immediately prior to `performCharacterChecks`. I also think the original comment is clearer, but you should keep your example, so try combining the two.

File: src/compiler/checker.ts (hunk 7 above)
Comment (yuit) [inline src/compiler/checker.ts]: :cactus:

File: src/compiler/checker.ts (hunk 3 above)
Comment (yuit) [inline src/compiler/checker.ts]: Thanks @JsonFreeman for explaining this. This is the logic so that if we didn't come from static member of class or interface, add the type parameters into the symbol table (type-parameters of class Declaration and interface are in member property of the symbol.
Note: that the memberFlags come from previous iteration

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Also note that for type aliases and signatures, the type parameters are in the locals, but for classes and interfaces, the type parameters are in the members.

File: src/harness/fourslash.ts (hunk 8 above)
Comment (yuit) [inline src/harness/fourslash.ts]: so for the completion list, I also want to make sure the following case
```typescript
class Foo {} -> additional parameter will be able to make sure that the list doesn't contain symbol with name Foo and of kind local class
var x = class Foo { /**/ } -> similar here but for kind class
```
Note: you can just pass the name if you don't kind the rest

File: src/services/services.ts (hunk 9 above)
Comment (yuit) [inline src/services/services.ts]: You can't use ```getDeclaredName```. For example: ```"/u0042"``` will be store on its symbol as "B" while if you use ```getDeclaredName``` it will get the unicode escape sequences which is as expected from the function since it is a declared name.

File: src/harness/fourslash.ts (hunk 8 above)
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Ah, I see. Got it

File: src/services/services.ts (hunk 9 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: Perhaps `getDeclaredName` should be amended then.

Comment (yuit) [inline src/services/services.ts]: I feel the same and I had some debate with myself about that. I also consider moving logic to check the external module name into ```getDeclaredName``` all together.

File: src/harness/fourslash.ts (hunk 8 above)
Comment (yuit) [inline src/harness/fourslash.ts]: Fix the comment at line 709

File: src/compiler/checker.ts (hunk 7 above)
Comment (yuit) [inline src/compiler/checker.ts]: Remove the old comment

File: src/services/services.ts (hunk 10 above)
Comment (yuit) [inline src/services/services.ts]: :cactus:

File: src/services/services.ts (hunk 11 above)
Comment (yuit) [inline src/services/services.ts]: :cactus:

File: src/services/services.ts (hunk 9 above)
Comment (yuit) [inline src/services/services.ts]: As per our discussion, there is another way to fix the case of classExpression and functionExpression which is to change bindAnonymouseDeclaration to bind classExpression or functionExpression to its declared name if it has one. I will experiment with that in a separate PR

File: src/harness/fourslash.ts (hunk 8 above)
Comment (yuit) [inline src/harness/fourslash.ts]: :cactus:

File: src/compiler/checker.ts (hunk 7 above)
Comment (yuit) [inline src/compiler/checker.ts]: :cactus:

File: src/services/services.ts (hunk 2 above)
Comment (yuit) [inline src/services/services.ts]: :cactus: file a bug

File: src/services/utilities.ts [hunk 12]
```diff
@@ -676,9 +676,16 @@ namespace ts {
...
+     * Strip off existed single quotes or double quotes from a given string
+     *
+     * @return non-quoted string
+     */
     export function stripQuotes(name: string) {
         let length = name.length;
-        if (length >= 2 && name.charCodeAt(0) === CharacterCodes.doubleQuote && name.charCodeAt(length - 1) === CharacterCodes.doubleQuote) {
+        if (length >= 2 &&
+            name.charCodeAt(0) === name.charCodeAt(length - 1) &&
+            (name.charCodeAt(0) === CharacterCodes.doubleQuote || name.charCodeAt(0) === CharacterCodes.singleQuote)) {
```
Comment (JsonFreeman) [inline src/services/utilities.ts]: Nice, that is cute :)

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -13371,11 +13371,6 @@ namespace ts {
                         case SyntaxKind.EnumDeclaration:
                             copySymbols(getSymbolOfNode(location).exports, meaning & SymbolFlags.EnumMember);
                             break;
-                        case SyntaxKind.ClassExpression:
-                            if ((<ClassExpression>location).name) {
-                                copySymbol(location.symbol, meaning);
-                            }
-                        // Fall through
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Don't you still need the fall through?

File: src/compiler/checker.ts [hunk 14]
```diff
@@ -13396,11 +13393,17 @@ namespace ts {
                 copySymbols(globals, meaning);
             }
 
-            // Returns 'true' if we should stop processing symbols.
-            function copySymbol(symbol: Symbol, meaning: SymbolFlags): void {
+            /**
+             * Copy the given symbol into symbol tables if the symbol has the given meaning
+             * and it doesn't already existed in the symbol table
+             * @param key a key for storing in symbol table; if null, use symbol.name
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: if undefined

File: src/compiler/checker.ts (hunk 13 above)
Comment (yuit) [inline src/compiler/checker.ts]: You don't need it. Break from the switch and fall through in this case behaves the same

File: src/compiler/checker.ts [hunk 15]
```diff
@@ -13396,11 +13393,17 @@ namespace ts {
...
+             * Copy the given symbol into symbol tables if the symbol has the given meaning
+             * and it doesn't already existed in the symbol table
+             * @param key a key for storing in symbol table; if null, use symbol.name
+             * @param symbol the symbol to be added into symbol table
+             * @param meaning meaning of symbol to filter by before adding to symbol table
+             */
+            function copySymbol(key: string, symbol: Symbol, meaning: SymbolFlags): void {
                 if (symbol.flags & meaning) {
-                    let id = symbol.name;
-                    if (!isReservedMemberName(id) && !hasProperty(symbols, id)) {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Why did you remove the reserved check?

File: src/compiler/checker.ts [hunk 16]
```diff
@@ -13396,11 +13393,17 @@ namespace ts {
...
-            // Returns 'true' if we should stop processing symbols.
-            function copySymbol(symbol: Symbol, meaning: SymbolFlags): void {
+            /**
+             * Copy the given symbol into symbol tables if the symbol has the given meaning
+             * and it doesn't already existed in the symbol table
+             * @param key a key for storing in symbol table; if null, use symbol.name
+             * @param symbol the symbol to be added into symbol table
+             * @param meaning meaning of symbol to filter by before adding to symbol table
+             */
+            function copySymbol(key: string, symbol: Symbol, meaning: SymbolFlags): void {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: You don't need the key here if you make that binder change we discussed

File: src/services/services.ts [hunk 17]
```diff
@@ -1424,6 +1424,9 @@ namespace ts {
         // class X {}
         export const classElement = "class";
 
+        // var x = class X {}
+        export const localClassElement = "local class";
```
Comment (JsonFreeman) [inline src/services/services.ts]: Why do we have a separate tag for this?

File: src/compiler/checker.ts (hunk 13 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]:
```ts
var c = class C<T> {
    prop: /*completion should show T*/
}
```

File: src/compiler/checker.ts (hunk 14 above)
Comment (yuit) [inline src/compiler/checker.ts]: :cactus:

File: src/compiler/checker.ts (hunk 16 above)
Comment (yuit) [inline src/compiler/checker.ts]: yep. I am doing it in separate PR :cactus:

File: src/compiler/checker.ts (hunk 15 above)
Comment (yuit) [inline src/compiler/checker.ts]: Because we will be doing it already when convert symbols to array (see ```symbolsToArray```)

File: src/services/utilities.ts (hunk 12 above)
Comment (yuit) [inline src/services/utilities.ts]: :cactus: :camel:

File: src/compiler/checker.ts (hunk 13 above)
Comment (yuit) [inline src/compiler/checker.ts]: :cactus:

File: src/services/services.ts (hunk 17 above)
Comment (yuit) [inline src/services/services.ts]: To be consistent with function expression which has kind of "local function"

Comment (JsonFreeman) [inline src/services/services.ts]: Not sure why we do it for either one then.

**Comment by yuit** [comment]
Fix @JsonFreeman  in newer commit

File: src/compiler/checker.ts [hunk 18]
```diff
@@ -13412,18 +13412,26 @@ namespace ts {
...
+                            // add the type parameters into the symbol table 
+                            // (type parameters of classDeclaration/classExpression and interface are in member property of the symbol.
+                            // Note: that the memberFlags come from previous iteration.
                             if (!(memberFlags & NodeFlags.Static)) {
                                 copySymbols(getSymbolOfNode(location).members, meaning & SymbolFlags.Type);
                             }
                             break;
                         case SyntaxKind.FunctionExpression:
-                            if ((<FunctionExpression>location).name) {
+                        case SyntaxKind.ClassExpression:
```
Comment (mhegazy) [inline src/compiler/checker.ts]: class expression listed twice

File: src/services/services.ts (hunk 17 above)
Comment (mhegazy) [inline src/services/services.ts]: i agree. let's take them both out.

File: src/compiler/checker.ts (hunk 18 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Uh oh. It'd be good to have a feature that catches this.

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: @JsonFreeman it's being tracked in #2854

**Comment by mhegazy** [comment]
:+1:  just remove the class expression label

**Comment by JsonFreeman** [comment]
@yuit Do you still need to do all these things now that you fixed the binder?

**Comment by yuit** [comment]
@JsonFreeman I make a change so that the service layer will get name through ```getDeclaredName``` which I prefer over using symbol.getName(). This makes completion list more consistent in the case of export default in that we won't show unicode escapes sequence in completion list at all. This will be some what consistent with other feature which show unicode escape sequence in its original declared form.

**Comment by JsonFreeman** [comment]
Great! Thanks Yui!

**Comment by JsonFreeman** [comment]
:+1:

**Comment by yuit** [comment]
@JsonFreeman  Thank you for the good discussion we had :smile:
